src/reference/add-wasm-support-to-crate.md

```diff
@@ -0,0 +1,156 @@
+your crate doesn't have any of those things, it likely already supports
+WebAssembly!
+
+You can always check by running `cargo build` for the WebAssembly target:
+
+```
+cargo build --target wasm32-unknown-unknown
+```
+
+If that command fails, then your crate doesn't support WebAssembly right now. If
+it doesn't fail, then your crate *might* support WebAssembly. You can be 100%
+sure that it does (and continues to do so!) by [adding tests for wasm and
+running those tests in CI.](#maintaining-ongoing-support-for-webassembly)
+
+## Adding Support for WebAssembly
+
+### Avoid Performing I/O Directly
+
+On the Web, I/O is always asynchronous, and there isn't a file system. Factor
+I/O out of your library, let users perform the I/O and then pass the input
+slices to your library instead.
+
+For example, refactor this:
+
+```rust
+use std::fs;
+use std::path::Path;
+
+pub fn parse_thing(path: &Path) -> Result<MyThing, MyError> {
+    let contents = fs::read(path)?;
+    // ...
+}
+```
+
+Into this:
+
+```rust
+pub fn parse_thing(contents: &[u8]) -> Result<MyThing, MyError> {
+    // ...
+}
+```
+
+### Add `wasm-bindgen` as a Dependency
+
+If you need to interact with the outside world (i.e. you can't have library
+consumers drive that interaction for you) then you'll need to add `wasm-bindgen`
+(and `js-sys` and `web-sys` if you need them) as a dependency for when
+compilation is targeting WebAssembly:
+
+```toml
+[target.'cfg(target_arch = "wasm32")'.dependencies]
+wasm-bindgen = "0.2"
+js-sys = "0.3"
+web-sys = "0.3"
+```
+
+### Avoid Synchronous I/O
+
+If you must perform I/O in your library, then it cannot be synchronous. There is
+only asynchronous I/O on the Web. Use [the `futures`
+crate](https://crates.io/crates/futures) and [the `wasm-bindgen-futures`
+crate](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen_futures/) to
+manage asynchronous I/O. If you're library functions are generic over some
+future type `F`, then that future can be implemented via `fetch` on the Web or
+via non-blocking I/O provided by the operating system.
+
+```rust
+pub fn do_stuff<F>(future: F) -> impl Future<Item = MyOtherThing>
+where
+    F: Future<Item = MyThing>,
+{
+    // ...
+}
+```
+
+You can also define a trait and implement it for WebAssembly and the Web and
+also for native targets:
+
+```rust
+trait ReadMyThing {
+    type F: Future<Item = MyThing>;
+    fn read(&self) -> Self::F;
+}
+
+#[cfg(target_arch = "wasm32")]
+struct WebReadMyThing {
+    // ...
+}
+
+#[cfg(target_arch = "wasm32")]
+impl ReadMyThing for WebReadMyThing {
+    // ...
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+struct NativeReadMyThing {
+    // ...
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl ReadMyThing for NativeReadMyThing {
+    // ...
+}
+```
+
+### Avoid Spawning Threads
+
+You can use `#[cfg(..)]`s to enable threaded and non-threaded code paths
+depending on if the target is WebAssembly or not:
```

Comment:
Another good thing to mention here is, like I/O, allow users to "bring their own threads". For example libraries, to be amenable with wasm, would expose the functions that are parallelize internally to allow users to externally parallelize them.

It may also be worth mentioning that one day in the far future threading may work, but the threads proposal (insert link here) isn't adding support to wasm to enable usage of `std::thread`